packages/api/endpoints/granules.js

```diff
@@ -331,6 +324,90 @@ async function put(req, res) {
     });
   }
   return res.boom.badRequest('Action is not supported. Choices are "applyWorkflow", "move", "reingest", "removeFromCmr" or specify no "action" to update an existing granule');
+};
+
+/**
+ * Update a single granule by granuleId and collectionId.
+ * Supported Actions: reingest, move, applyWorkflow, RemoveFromCMR.
+ * If no action is included on the request, the body is assumed to be an
+ * existing granule to update, and update is called with the input parameters.
+ *
+ * @param {Object} req - express request object
+ * @param {Object} res - express response object
+ * @returns {Promise<Object>} the promise of express response object
+ */
+async function put(req, res) {
+  const {
+    knex = await getKnexClient(),
+    granulePgModel = new GranulePgModel(),
+    collectionPgModel = new CollectionPgModel(),
+  } = req.testContext || {};
+
+  if (!req.body.action) {
+    if (req.body.granuleId === req.params.granuleName) {
+      return putGranule(req, res);
+    }
```

Comment:
We should check if the body.collectionId matches the params.collectionId.  I tried to update a granule with different collectionId in the body, it ends up two granules in the postgres.
curl --request PUT https://dz5bdrhmhd.execute-api.us-east-1.amazonaws.com:8000/dev/granules/MOD09GQ___006/MOD09GQ.A0638263._eksMj.006.2933852732093 \
--header 'Authorization: Bearer replaceWithToken' \
--header 'Content-Type: application/json' \
  --data '{
  "granuleId": "MOD09GQ.A0638263._eksMj.006.2933852732093", "collectionId": "MOD09GQjl-tf-DeadLetterArchiveProcessing-1625581933440___006",
"duration": 1000,
  "status": "completed"
  }'
{"message":"Successfully wrote granule with Granule Id: MOD09GQ.A0638263._eksMj.006.2933852732093, Collection Id: MOD09GQjl-tf-DeadLetterArchiveProcessing-1625581933440___006"}

"cumulus_id"	"granule_id"	"status"	"collection_cumulus_id"	"created_at"	"updated_at"	"published"	"duration"	"time_to_archive"	"time_to_process"	"product_volume"	"error"	"cmr_link"	"pdr_cumulus_id"	"provider_cumulus_id"	"beginning_date_time"	"ending_date_time"	"last_update_date_time"	"processing_end_date_time"	"processing_start_date_time"	"production_date_time"	"query_fields"	"timestamp"
17491	"MOD09GQ.A0638263._eksMj.006.2933852732093"	"completed"	692	"2022-07-26 18:48:59.001+00"	"2022-07-26 18:48:59.001+00"	false	1000				"{}"											"2022-07-26 18:48:59.001+00"
16494	"MOD09GQ.A0638263._eksMj.006.2933852732093"	"completed"	1124	"2022-06-15 18:20:38.522+00"	"2022-06-15 18:20:45.832+00"	false	7.31	0	0.398	2348	"{""Cause"": ""None"", ""Error"": ""Unknown Error""}"			2477	"2017-10-24 00:00:00+00"	"2017-11-08 23:59:59+00"	"2018-04-25 21:45:45.524+00"	"2022-06-15 18:20:44.912+00"	"2022-06-15 18:20:39.168+00"	"2018-07-19 12:01:01+00"		"2022-06-15 18:20:45.832+00"

But there is only one showing up in the dashboard.
<img width="1792" alt="Screen Shot 2022-07-25 at 1 27 15 PM" src="https://user-images.githubusercontent.com/34660846/181088954-752643aa-7a65-4fe0-b6cf-f63109b3b239.png">

Reply:
I also retried to create a granule with a different collectionId from existing record.  Existing granule: MOD09GQ.A0550661.rjQgx7.006.1139022551166 collectionId MOD09GQ___006

Post request:
{
    "granuleId": "MOD09GQ.A0550661.rjQgx7.006.1139022551166",
    "collectionId": "MOD09GQjl-tf-DeadLetterArchiveProcessing-1625581933440___006",
    "status": "running"
}
Result in two granules in postgres, they have different contents.
"cumulus_id"	"granule_id"	"status"	"collection_cumulus_id"	"created_at"	"updated_at"	"published"	"duration"	"time_to_archive"	"time_to_process"	"product_volume"	"error"	"cmr_link"	"pdr_cumulus_id"	"provider_cumulus_id"	"beginning_date_time"	"ending_date_time"	"last_update_date_time"	"processing_end_date_time"	"processing_start_date_time"	"production_date_time"	"query_fields"	  ##"timestamp"
17492	"MOD09GQ.A0550661.rjQgx7.006.1139022551166"	"running"	692	"2022-07-26 20:09:35.087+00"	"2022-07-26 20:09:35.087+00"	false					"{}"											"2022-07-26 20:09:35.087+00"
16068	"MOD09GQ.A0550661.rjQgx7.006.1139022551166"	"completed"	1124	"2022-06-15 18:20:18.032+00"	"2022-06-15 18:20:24.933+00"	false	6.901	0	0.373	2348	"{""Cause"": ""None"", ""Error"": ""Unknown Error""}"			2477	"2017-10-24 00:00:00+00"	"2017-11-08 23:59:59+00"	"2018-04-25 21:45:45.524+00"	"2022-06-15 18:20:24.061+00"	"2022-06-15 18:20:18.136+00"	"2018-07-19 12:01:01+00"		"2022-06-15 18:20:24.933+00"
One record in elasticsearch, it has the combined information with files information get lost.
{
    "beginningDateTime": "2017-10-24T00:00:00.000Z",
    "collectionId": "MOD09GQjl-tf-DeadLetterArchiveProcessing-1625581933440___006",
    "createdAt": 1658866175087,
    "duration": 6.901,
    "endingDateTime": "2017-11-08T23:59:59.000Z",
    "error": {},
    "execution": "https://console.aws.amazon.com/states/home?region=us-east-1#/executions/details/arn:aws:states:us-east-1:12345:execution:jl-tf-IngestGranule:bc0a274b-d3d8-4e3f-906d-8924634428fd",
    "files": [],
    "granuleId": "MOD09GQ.A0550661.rjQgx7.006.1139022551166",
    "lastUpdateDateTime": "2018-04-25T21:45:45.524Z",
    "processingEndDateTime": "2022-06-15T18:20:24.061Z",
    "processingStartDateTime": "2022-06-15T18:20:18.136Z",
    "productionDateTime": "2018-07-19T12:01:01.000Z",
    "productVolume": "2348",
    "provider": "s3provider",
    "published": false,
    "status": "running",
    "timestamp": 1658866175087,
    "timeToArchive": 0,
    "timeToPreprocess": 0.373,
    "updatedAt": 1658866175087
  }

Reply:
I think the current elasticsearch can't handle two granules with the same granuleId but different collectionId.  We need to address this or separate ticket.

Reply:
Also dashboard can't view the updated granule if there are multiple granules with same granuleId and different collectionId in the postgres.  The error in the dashboard: Failed to write MOD09GQ.A0550661.rjQgx7.006.1139022551166 due to granuleId duplication on postgres granule record
I think it's expected behavior, and dashboard should get granule with granuleId + collectionId in the future.

Reply:
A partial answer: 

I made a new ticket for the ElasticSearch updates to support multiple granules with the same granuleIds. You're right that this is a major concern and something we haven't considered: https://bugs.earthdata.nasa.gov/browse/CUMULUS-3004

There is also an existing ticket for the dashboard updates that will support all of these /collectionId/granuleId endpoint updates: https://bugs.earthdata.nasa.gov/browse/CUMULUS-2982. If this API ticket is any indication however, we may want to split the dashboard ticket up.

I've made the change to check the collectionId queryparam against the body. Once the integration test changes are sorted I will link that here.

Reply:
One more phase 3 ticket to update all integration tests to use these new endpoints: https://bugs.earthdata.nasa.gov/browse/CUMULUS-3015

Note that a few integration tests were already switched over to the new endpoints in this PR so that we can at least broadly test the changes: https://github.com/nasa/cumulus/pull/3022/files#diff-43db1f9ffc606dcbab3ac394871cd5fa8229533021159a9e55d918a1bf656abb